File: Linked_List/143.Reorder-List/143.Reorder-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """        
        if not head: return None
        
        fast,slow = head,head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        head2 = slow.next
        slow.next = None
                  
        prv,h = None,head2
        while h:
            nxt = h.next
            h.next = prv
            prv = h
            h = nxt
        head2 = prv       
            
        h = ListNode(0)
        h1,h2 = head,head2
        while h1 and h2:
            h1_temp, h2_temp = h1.next, h2.next
            h.next = h1     
            h = h.next
            h.next = h2
            h = h.next
            h1,h2 = h1_temp, h2_temp
        if h1:
            h.next = h1

File: Linked_List/143.Reorder-List/143.Reorder-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) 
    {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        ListNode* p = dummy;
        int count = 0;
        while (p->next)
        {
            count++;
            p = p->next;
        }

        ListNode* q = dummy;
        for (int i=0; i<(count+1)/2; i++)
            q = q->next;
        ListNode* head2 = q->next;
        q->next = NULL;

        head2 = reverseLinkedList(head2);

        p = head, q = head2;
        ListNode* h = dummy;
        while (p || q)
        {
            if (p)
            {
                h->next = p;
                p = p->next;
                h = h->next;
            }
            if (q)
            {
                h->next = q;
                q = q->next;
                h = h->next;
            }
        }
    }

    ListNode* reverseLinkedList(ListNode* head)
    {
        ListNode* cur = head;
        ListNode* last = NULL;
        ListNode* next = NULL;
        while (cur!=NULL)
        {
            next = cur->next;
            cur->next = last;
            last = cur;
            cur = next;
        }
        return last;
    }
};
